### PR TITLE
Initial commit of audio/I2S driver without of any BSP support for any…

### PR DESCRIPTION
**Pull request by btheosam**
… hardware

### CODE REVIEW

**Comment by btheosam** [comment]
This review request has comments in continuation to previous comments by sunghan-chang but I had to delete those requests twice due to my fault of corrupting those branches.

File: os/drivers/audio/audio_null.c [hunk 1]
```diff
@@ -0,0 +1,811 @@
...
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct null_dev_s {
+	struct audio_lowerhalf_s dev;	/* Audio lower half (this device) */
+	mqd_t mq;					/* Message queue for receiving messages */
+	char mqname[16];			/* Our message queue name */
+	pthread_t threadid;			/* ID of our thread */
+#ifndef CONFIG_AUDIO_EXCLUDE_STOP
```
Comment (sunghan-chang) [inline os/drivers/audio/audio_null.c]: There is no AUDIO_EXCLUDE_STOP on Kconfig.

File: os/drivers/audio/audio_null.c [hunk 2]
```diff
@@ -0,0 +1,811 @@
...
+	volatile bool terminate;	/* True: request to terminate */
+#endif
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int null_getcaps(FAR struct audio_lowerhalf_s *dev, int type, FAR struct audio_caps_s *caps);
+#ifdef CONFIG_AUDIO_MULTI_SESSION
```
Comment (sunghan-chang) [inline os/drivers/audio/audio_null.c]: Same as previous

File: os/drivers/audio/audio_null.c [hunk 3]
```diff
@@ -0,0 +1,811 @@
...
+static int null_start(FAR struct audio_lowerhalf_s *dev);
+#endif
+#ifndef CONFIG_AUDIO_EXCLUDE_STOP
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+static int null_stop(FAR struct audio_lowerhalf_s *dev, FAR void *session);
+#else
+static int null_stop(FAR struct audio_lowerhalf_s *dev);
+#endif
+#endif
+#ifndef CONFIG_AUDIO_EXCLUDE_PAUSE_RESUME
```
Comment (sunghan-chang) [inline os/drivers/audio/audio_null.c]: I think all of config in this file are missed.

File: os/drivers/audio/audio_null.c [hunk 4]
```diff
@@ -0,0 +1,811 @@
...
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct null_dev_s {
+	struct audio_lowerhalf_s dev;	/* Audio lower half (this device) */
+	mqd_t mq;					/* Message queue for receiving messages */
```
Comment (sunghan-chang) [inline os/drivers/audio/audio_null.c]: Don't you want to align for comments?

**Comment by btheosam** [comment]
Thanks for the review, the audio framework is not yet ready to be pushed along. Hence removed all audio related driver and test applcations. This commit has only I2S driver and I2S char device (may be used to test).

File: os/include/tinyara/audio/audio.h [hunk 5]
```diff
@@ -0,0 +1,727 @@
...
+#define AUDIO_APB_DEQUEUED          (1 << 2)
+#define AUDIO_APB_FINAL             (1 << 3)	/* Last buffer in the stream */
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+/* Define the size of AP Buffer sample count base on CONFIG */
+
+#ifdef CONFIG_AUDIO_LARGE_BUFFERS
```
Comment (btheosam) [inline os/include/tinyara/audio/audio.h]: Many of these configuration paramentes should be from audio framework layer.

Comment (btheosam) [inline os/include/tinyara/audio/audio.h]: Unless audio framework layer is developed these configuration parameters wont appear in Kconfig. Then why is it needed now? There is a strong dependency with I2S device, we need this to develop I2S for now.

**Pull request merged**